File: c/binary-search-tree/binary_search_tree.c
#include "binary_search_tree.h"

#include <stdlib.h>
#include <string.h>

node_t *insert(node_t *root, int data);
int tree_size(node_t *tree);

node_t *build_tree(int *tree_data, size_t tree_data_len) {

    node_t *root = NULL;

    for (size_t i = 0; i < tree_data_len; i++) {
        root = insert(root, tree_data[i]);
    }

    return root;
}


node_t *insert(node_t *root, int data) {
    
    if (!root) {
        node_t *new_node = malloc(sizeof (node_t));
        new_node->data = data;
        new_node->left = NULL;
        new_node->right = NULL;

        return new_node;
    }

    if (data <= root->data) {
        root->left = insert(root->left, data);
    }

    if (data > root->data) {
        root->right = insert(root->right, data);
    }

    return root;
}

void free_tree(node_t *tree) {

    if (tree->left == NULL && tree->right == NULL) {
        free(tree);
        return;
    }

    if (tree->left) {
        free_tree(tree->left);
    }

    if (tree->right) {
        free_tree(tree->right);
    }

}


int *sorted_data(node_t *tree) {

    if (!tree) {
        return NULL;
    }

    int* left_data = sorted_data(tree->left);
    int left_size = tree_size(tree->left);

    int* current = &(tree->data);

    int* right_data = sorted_data(tree->right);
    int right_size = tree_size(tree->right);

    int total_size = left_size + 1 + right_size;

    int* total_data = malloc(total_size * sizeof(int));

    memcpy(total_data, left_data, left_size * sizeof(int));
    memcpy(total_data + left_size, current, 1 * sizeof(int));
    memcpy(total_data + (left_size + 1), right_data, right_size * sizeof(int));

    return total_data;

}

int tree_size(node_t *tree) {

    if (!tree) {
        return 0;
    }

    return tree_size(tree->left) + 1 + tree_size(tree->right);
}

File: python/binary-search-tree/binary_search_tree.py
class TreeNode:
    def __init__(self, data, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    def __str__(self):
        return f'TreeNode(data={self.data}, left={self.left}, right={self.right})'


class BinarySearchTree:
    def __init__(self, tree_data):

        self.root = None

        for item in tree_data:
            self.root = BinarySearchTree.insert_data(self.root, item)

    @staticmethod
    def insert_data(current_node, data):

        if not current_node:
            return TreeNode(data, None, None)

        if int(data) <= int(current_node.data):
            return TreeNode(
                current_node.data,
                BinarySearchTree.insert_data(current_node.left, data),
                current_node.right
            )

        return TreeNode(
            current_node.data,
            current_node.left,
            BinarySearchTree.insert_data(current_node.right, data)
        )

    def data(self):
        return self.root

    def sorted_data(self):
        return BinarySearchTree.get_sorted_data(self.root)

    @staticmethod
    def get_sorted_data(root_node):
        if not root_node:
            return []

        return BinarySearchTree.get_sorted_data(root_node.left) + \
            [root_node.data] + \
            BinarySearchTree.get_sorted_data(root_node.right)
